File: 2018/python/day03.py
def parse(str):
    issue, vals = str.split(" @ ")
    coord, size = vals.split(": ")
    issue = int(issue.replace("#", ""))
    coord = [int(i) for i in coord.split(",")]
    size = [int(i) for i in size.split("x")]
    return (issue, coord, size)


file = open("data/input-03.txt")
vals = [parse(i) for i in file.readlines()]

cover, claim = {}, {}
for (issue, coord, size) in vals:
    lef, top = coord
    wid, hei = size
    for x in range(lef, lef + wid):
        for y in range(top, top + hei):
            val = (x, y)
            if val in cover.keys():
                cover[val] += 1
                claim[val].append(issue)
            else:
                cover[val] = 1
                claim[val] = [issue]

res = sum([i > 1 for i in cover.values()])
print(f"Part 1: {res}")
assert res == 111326

used = [j for i, j in zip(cover.values(), claim.values()) if i > 1]
used = set([i for inner in used for i in inner])  # unlist and unique
[left] = list(set([issue for issue, _, _ in vals]).difference(used))
print(f"Part 2: {left}")
assert left == 1019

File: 2018/julia/day03.jl
function parseLine(str)
    issue, vals = split(str, " @ ")
    coord, size = split(vals, ": ")
    issue = parse(Int, replace(issue, "#" => ""))
    coord = [parse(Int, i) for i in split(coord, ",")]
    size  = [parse(Int, i) for i in split(size, "x")]
    issue, coord, size
end

vals = [parseLine(i) for i in readlines("data/input-03.txt")]
cover, claim = Dict(), Dict()
for (issue, coord, size) in vals
    lef, top = coord
    wid, hei = size
    for x in lef:lef + wid - 1
        for y in top:top + hei - 1
            val = (x, y)
            if val in keys(cover)
                global cover[val] += 1
                push!(claim[val], issue)
            else
                global cover[val] = 1
                global claim[val] = [issue]
            end
        end
    end
end

res = sum([i > 1 for i in values(cover)])
println("Part 1: $(res)")
@assert(res == 111326)

used = [j for (i, j) in zip(values(cover), values(claim)) if i > 1]
used = Set([i for inner in used for i in inner])
left = collect(setdiff(Set([issue for (issue, _, _) in vals]), used))[1]
println("Part 2: $(left)")
@assert(left == 1019)
